resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala

```diff
@@ -320,50 +322,83 @@ private[spark] class KubernetesClusterSchedulerBackend(
     override def eventReceived(action: Action, pod: Pod): Unit = {
       val podName = pod.getMetadata.getName
       val podIP = pod.getStatus.getPodIP
-
+      val podPhase = pod.getStatus.getPhase
       action match {
-        case Action.MODIFIED if (pod.getStatus.getPhase == "Running"
+        case Action.MODIFIED if (podPhase == "Running"
             && pod.getMetadata.getDeletionTimestamp == null) =>
           val clusterNodeName = pod.getSpec.getNodeName
           logInfo(s"Executor pod $podName ready, launched at $clusterNodeName as IP $podIP.")
           executorPodsByIPs.put(podIP, pod)
 
-        case Action.DELETED | Action.ERROR =>
+        case Action.MODIFIED if (podPhase == "Init:Error" || podPhase == "Init:CrashLoopBackoff")
+          && pod.getMetadata.getDeletionTimestamp == null =>
           val executorId = getExecutorId(pod)
-          logDebug(s"Executor pod $podName at IP $podIP was at $action.")
-          if (podIP != null) {
-            executorPodsByIPs.remove(podIP)
+          failedInitExecutors.add(executorId)
+          if (failedInitExecutors.size >= executorMaxInitErrors) {
+            val errorMessage = s"Aborting Spark application because $executorMaxInitErrors" +
+              s" executors failed to start. The maximum number of allowed startup failures is" +
+              s" $executorMaxInitErrors. Please contact your cluster administrator or increase" +
+              s" your setting of ${KUBERNETES_EXECUTOR_MAX_INIT_ERRORS.key}."
+            logError(errorMessage)
+            KubernetesClusterSchedulerBackend.this.scheduler.sc.stopInNewThread()
```

Comment:
Why you need to call this?

Reply:
The problem is that executors that fail to start up won't create task failures, meaning that they will not cause the job(s) to stop. Additionally, executors failing to start up will affect every job on the process. So what I think we want to do is when we fail with these initialization errors too many times, is to shut down the Spark context and crash all of the jobs. The alternative is that we stop requesting new executors entirely, but this state isn't desirable because for example you could have your Spark application running with 0 executors, meaning that your application can be stuck for seemingly no reason.

Reply:
Even running at degraded can be concerning though. What if we only get 1 executor for an application that's processing terabytes of data? I don't think there's an intuitive default here, though perhaps if we're at say 75% of requested capacity or some arbitrary number then we can continue. Alternatively, we can just retry requesting executors forever and never consider these executor startup failures to be real problems (which would remove this if statement, for example). I don't think it's entirely unreasonable to always retry requesting executors when they fail to start up.

Reply:
Always re-requesting executors has the potential to overwhelm etcd with failed pods especially for long running jobs. This seems overly conservative but a reasonable place to start - fail entire task upon N failures. With dynamic allocation, it should be possible to check that at least `spark.dynamicAllocation.minExecutors` are alive and make a decision accordingly.

Reply:
Leaning on configured minimum executors could be construed as a way to let the user express application-dependent throughput minimums. If the use case is terabyte-scale, then setting minimum executors appropriately would address that.  A %-threshold is also reasonable, it just adds a new knob.

Requesting new executors indefinitely seems plausible, as long as failed pods don't accumulate. Having a submission hang indefinitely while kube churns may be confusing behavior, at least if users are not familiar with kube and hoping to treat it transparently. Maybe mesos backend policies could provide an analogy?

Reply:
@foxish out of curiosity how does replication controller and replica set, and similar controllers, handle the same problem? Presumably there would also be heavy etcd load in the case pods always fail to start up and the controller keeps trying to do so.

Reply:
The controllers have rate-limiting queues with exponential backoff. In the past, we've had issues (https://github.com/kubernetes/kubernetes/issues/30628, https://github.com/kubernetes/kubernetes/issues/27634 and many more) where a misconfigured queue has caused controllers to spew pods and retry.